cpp/include/kvikio/remote_handle.hpp

```diff
@@ -0,0 +1,356 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <cstdlib>
+#include <iostream>
+#include <memory>
+#include <stdexcept>
+#include <string>
+#include <utility>
+
+#include <aws/core/Aws.h>
+#include <aws/core/auth/AWSCredentialsProviderChain.h>
+#include <aws/core/utils/stream/PreallocatedStreamBuf.h>
+#include <aws/s3/S3Client.h>
+#include <aws/s3/model/GetObjectRequest.h>
+#include <aws/s3/model/HeadObjectRequest.h>
+
+#include <kvikio/parallel_operation.hpp>
+#include <kvikio/posix_io.hpp>
+#include <kvikio/utils.hpp>
+
+#ifndef KVIKIO_AWS_SDK_FOUND
+#error "cannot include <kvikio/remote_handle.hpp>, configuration did not find AWS SDK"
+#endif
+
+namespace kvikio {
+namespace detail {
+
+/**
+ * Stream implementation of a fixed size buffer.
+ */
+class BufferAsStream : public Aws::IOStream {
+ public:
+  using Base = Aws::IOStream;
+  explicit BufferAsStream(std::streambuf* buf) : Base(buf) {}
+
+  ~BufferAsStream() override = default;
+};
+
+/**
+ * @brief Given a file path like "s3://<bucket>/<object>", return the name of the bucket and object.
+ *
+ * @param path S3 file path.
+ * @return Pair of strings: [bucket-name, object-name].
+ */
+inline std::pair<std::string, std::string> parse_s3_path(const std::string& path)
+{
+  if (path.empty()) { throw std::invalid_argument("The remote path cannot be an empty string."); }
+  if (path.size() < 5 || path.substr(0, 5) != "s3://") {
+    throw std::invalid_argument("The remote path must start with the S3 scheme (\"s3://\").");
+  }
+  std::string p = path.substr(5);
+  if (p.empty()) { throw std::invalid_argument("The remote path cannot be an empty string."); }
+  size_t pos = p.find_first_of('/');
+  if (pos == 0) { throw std::invalid_argument("The remote path does not contain a bucket name."); }
+  return std::make_pair(p.substr(0, pos), (pos == std::string::npos) ? "" : p.substr(pos + 1));
+}
+
+}  // namespace detail
+
+/**
+ * @brief S3 context, which initializes and maintains the S3 SDK and client.
+ *
+ * If not given an existing S3 client, S3Context calls `Aws::InitAPI()` and `Aws::ShutdownAPI`,
+ * which inherit some limitations from the SDK.
+ *  - The SDK for C++ and its dependencies use C++ static objects, and the order of static object
+ *    destruction is not determined by the C++ standard. To avoid memory issues caused by the
+ *    nondeterministic order of static variable destruction, do not wrap `S3Context` in another
+ *    static object.
+ *  - Please construct and destruct `S3Context` from the same thread (use a dedicated thread if
+ *    necessary). This avoids problems in initializing the dependent Common RunTime C libraries.
+ */
+class S3Context {
+ private:
+  // We use a shared point since constructing a default `Aws::S3::S3Client` before calling
+  // `Aws::InitAPI` is illegal.
+  std::shared_ptr<Aws::S3::S3Client> _client;
```

Comment:
I guess we also need an owning pointer because in the case where we construct the S3Context from an endpoint, we need to keep the client alive.

Reply:
yes